### PR TITLE
remove duplication of content when translating (WP-772)

### PR DESCRIPTION
**Pull request by vsolovei-smartling**

### CODE REVIEW

File: inc/Smartling/ContentTypes/ConfigParsers/FieldFilterConfigParser.php [hunk 1]
```diff
@@ -268,39 +272,31 @@ private function validateRelatedType()
...
-     */
     public function __construct(array $config, ContainerBuilder $di)
     {
         $this->setRawConfig($config);
         $this->setDi($di);
+        $submissionManager = $this->getService('manager.submission');
+        if (!$submissionManager instanceof SubmissionManager) {
+            throw new SmartlingConfigException(SubmissionManager::class . ' expected in DI for `manager.submission`');
+        }
+        $this->submissionManager = $submissionManager;
```
Comment (PavelLoparev) [inline inc/Smartling/ContentTypes/ConfigParsers/FieldFilterConfigParser.php:290]: Why don't we inject service directly?

File: inc/Smartling/ContentTypes/ConfigParsers/FieldFilterConfigParser.php [hunk 2]
```diff
@@ -268,39 +272,31 @@ private function validateRelatedType()
...
-     * @return SerializerInterface
-     */
-    private function getSerializerInstance()
+    private function getSerializerInstance(): SerializerInterface
     {
         return $this->getService('manager.serializer')->getSerializer($this->getSerialization());
     }
 
-    private function getLocalizeFilter()
+    private function getLocalizeFilter(): CustomTypeProcessor|TermBasedProcessor|MediaBasedProcessor|PostBasedProcessor
```
Comment (PavelLoparev) [inline inc/Smartling/ContentTypes/ConfigParsers/FieldFilterConfigParser.php:299]: Should be typed with parent class?

File: inc/Smartling/Submissions/SubmissionEntity.php [hunk 3]
```diff
@@ -460,45 +452,52 @@ public function setSourceId(int $source_id): SubmissionEntity
...
+        return $this;
     }
 
-    protected function setFileUri(?string $file_uri): SubmissionEntity
+    /**
+     * @throws SmartlingDirectRunRuntimeException
+     * @throws SmartlingInvalidFactoryArgumentException
+     * @throws BlogNotFoundException
+     */
+    public function generateFileUri(): self
```
Comment (PavelLoparev) [inline inc/Smartling/Submissions/SubmissionEntity.php:482]: Just wondering is there any difference between typing `self` vs explicit `SubmissionEntity`

File: inc/Smartling/Submissions/SubmissionManager.php [hunk 4]
```diff
@@ -97,6 +97,26 @@ public function submissionExistsNoLastError(string $contentType, int $sourceBlog
...
+        ], 2);
+        switch (count($submissions)) {
+            case 0:
+                return null;
+            case 2:
+                $this->getLogger()->error("More than one submissions exist, contentType=$contentType, sourceBlogId=$sourceBlogId, sourceId=$sourceId, targetBlogId=$targetBlogId");
+                // no break
+            default:
+                return ArrayHelper::first($submissions);
+        }
```
Comment (PavelLoparev) [inline inc/Smartling/Submissions/SubmissionManager.php:117]: You will not have log record if more than 2 rows returned?

File: inc/Smartling/Base/SmartlingCoreExportApi.php [hunk 5]
```diff
@@ -40,21 +41,19 @@ public function sendAttachmentForTranslation(int $sourceBlogId, int $targetBlogI
     }
 
     /**
-     * @param SubmissionEntity $submission
-     * @return string
+     * @throws SmartlingWpDataIntegrityException
      */
-    public function getAttachmentRelativePathBySubmission(SubmissionEntity $submission)
+    public function getAttachmentRelativePathBySubmission(SubmissionEntity $submission): string
```
[the remark this answers is not in the document]
Comment (vsolovei-smartling) [reply inc/Smartling/Base/SmartlingCoreExportApi.php:46]: Yes, it'll throw a TypeError. That said, we rely on it being a string in some places.

File: inc/Smartling/Submissions/SubmissionEntity.php [hunk 6]
```diff
@@ -460,45 +452,52 @@ public function setSourceId(int $source_id): SubmissionEntity
...
-                ->setSourceContentId($this->getSourceId());
-
-            $filterParams = apply_filters(ExportedAPI::FILTER_SMARTLING_FILE_URI, $filterParams);
-
-            if (($filterParams instanceof SmartlingFileUriFilterParamater)
-                && !StringHelper::isNullOrEmpty($filterParams->getFileUri())
-            ) {
-                $fileUri = $filterParams->getFileUri();
+            /** @noinspection JsonEncodingApiUsageInspection, it's ok to have no backtrace, this is for debugging to untangle this function into proper get and set pair */
+            $this->logger->warning("Set file uri on get, backtrace: " . json_encode(debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS)));
```
[the remark this answers is not in the document]
Comment (vsolovei-smartling) [reply inc/Smartling/Submissions/SubmissionEntity.php:459]: I wanted to make sure that no instances of this log appear, I did a lot of testing, but still might have missed something, I'd like this to stay

File: inc/Smartling/Submissions/SubmissionManager.php [hunk 7]
```diff
@@ -97,6 +97,26 @@ public function submissionExistsNoLastError(string $contentType, int $sourceBlog
             ], 1));
     }
 
+    public function findTargetSubmission(string $contentType, int $sourceBlogId, int $sourceId, int $targetBlogId): ?SubmissionEntity
```
[the remark this answers is not in the document]
Comment (vsolovei-smartling) [reply inc/Smartling/Submissions/SubmissionManager.php]: Removed, there already is a findOne

File: inc/Smartling/ContentTypes/ConfigParsers/FieldFilterConfigParser.php (hunk 1 above)
Comment (vsolovei-smartling) [reply inc/Smartling/ContentTypes/ConfigParsers/FieldFilterConfigParser.php:290]: It already has the entire DI Container, and refactoring it to inject service directly would require moving this code into Bootstrap.

File: inc/Smartling/ContentTypes/ConfigParsers/FieldFilterConfigParser.php (hunk 2 above)
Comment (vsolovei-smartling) [reply inc/Smartling/ContentTypes/ConfigParsers/FieldFilterConfigParser.php]: Done

File: inc/Smartling/Submissions/SubmissionEntity.php (hunk 3 above)
Comment (vsolovei-smartling) [reply inc/Smartling/Submissions/SubmissionEntity.php:482]: No difference

File: inc/Smartling/Submissions/SubmissionManager.php (hunk 4 above)
Comment (vsolovei-smartling) [reply inc/Smartling/Submissions/SubmissionManager.php]: This is obsolete, but there's a limit of 2 when searching, similar to findOne method, if more than one is found, it's two.

**Review by PavelLoparev — APPROVED** [review]

**Pull request merged**